### PR TITLE
Notifications: add mark all read button to nav bar

### PR DESCRIPTION
**Pull request by alpavanoglu**
Ref: #17135 

With this PR, the following 2 steps are addressed. PR's will follow for the remaining 4:
- [x] Add feature flag.
- [x] Add button to nav bar.
- [ ] Mark notifications read for the selected filter.
- [ ] Add tracks for feature usage.
- [ ] Accessibility support.
- [ ] Enable feature flag and add release note (after branch is merged to develop).

To Test:
1. Enable the Feature Flag: Settings -> AppSettings -> Debug -> Enable "Mark Notifications As Read"
2. Notifications Tab.
3. A Checkmark button must have been placed left of the Settings Icon. The button will be non-functional until the follow up PR is merged.
4. Reader Tab's Bar Buttons' positions are also adjusted to match the Design Guidelines. They all have 44x44 tappable area now and should match in positions between the two tabs (Reader-Notifications).

## Regression Notes
1. Potential unintended areas of impact
The "Quick Start" feature displays an indicator on top of the button. I do not know how to trigger this case and I must test it. Some guidance/steps here would be very much appreciated.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
TODO, see step 1.

3. What automated tests I added (or what prevented me from doing so)
I couldn't find a UI Test file for `NotificationsViewController.swift`. Please let me know if I should write one.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Screen Shot for Reference:
Screenshot 2022-01-15 at 16 44 01

### CODE REVIEW

File: WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift [hunk 1]
```diff
@@ -48,7 +49,6 @@ enum FeatureFlag: Int, CaseIterable, OverrideableFlag {
             return true
         case .siteIconCreator:
             return BuildConfiguration.current != .appStore
-            return true
```
Comment (alpavanoglu) [inline WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift:51]: I have removed this as it was raising a warning and would never be executed.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift [hunk 2]
```diff
@@ -92,10 +92,28 @@ class NotificationsViewController: UITableViewController, UIViewControllerRestor
...
+    lazy var markAllAsReadButton: UIButton = {
+        let markButton: UIButton = UIButton.init(type: .custom)
+        markButton.setImage(.gridicon(.checkmark), for: .normal)
+        markButton.addTarget(self, action: #selector(markAllAsRead), for: .touchUpInside)
+        markButton.translatesAutoresizingMaskIntoConstraints = false
+        markButton.contentMode = .right
+        markButton.accessibilityLabel = NSLocalizedString(
+            "Mark All As Read",
+            comment: "Marks all notifications under the filter as read"
+        )
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:120]: Where can I find the localizations? I will also need the strings to display a TOAST/NoticeView upon completion of marking the notifications.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift [hunk 3]
```diff
@@ -1658,7 +1691,6 @@ private extension NotificationsViewController {
         }
 
         static let sortKey = "timestamp"
-        static let allFilters = [Filter.none, .unread, .comment, .follow, .like]
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:1661]: This had only callsite and I thought CaseIterable does exactly this. I can revert the change if there are concerns.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        setupConstraints(for: rightControl)
+    }
+
+    private static func setupConstraints(for control: UIControl?) {
+        guard let control = control else {
+            return
+        }
+
+        // TODO: Check if a common/global constant exists
+        let minimumTappableLength: CGFloat = 44
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift]: If there's a common constant for this one I can remove this local constant and use that instead.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import UIKit
+
+// Where to place this file?
```
Comment (alpavanoglu) [inline WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift]: I am unfamiliar with the general folder structure so I placed it under `Tab Navigation` which is a helper adding for buttons to navigation bar. It is used from `NotificationsViewController` & `ReaderTabViewController`

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift (hunk 2 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:120]: Just following up here - as discussed in Slack, simply using NSLocalizedString is enough. The text will get translated later.

File: WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift (hunk 5 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift]: > Where to place this file?

Before I answer that, I'm a bit confused by the purpose of this file. I could be misunderstanding what this does, and a bit of a disclaimer - I've not been able to actually build and run this branch yet (my problem, not yours), so I've not been able to test it. So please forgive me if I got it wrong. But it seems to override the built-in navigation bar functionality of setting multiple right bar buttons (ex: `navigationItem.rightBarButtonItems = [searchBarButton, settingsButton]`) for the sole purpose of setting the buttons' width and height to 44. Could this not be done on each button before adding it to the nav bar? It seems to me sizing the buttons correctly is preferable to creating a whole custom stack view.

Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift]: No you're totally right. It only helped me avoid the code duplication there. The stack view grants us the control over the spacing between the bar button items. When I directly add `UIBarButtonItem` via `setBarButtonItems` function, there is a spacing between them that doesn't follow the current design. Another option I could think of was to use the appearance API to set the spacing to 0.
`UIStackView.appearance(whenContainedInInstancesOf: [UINavigationBar.self]).spacing`
I thought this is a bit too intrusive for me to use without communicating with the team. 

Another option I've found on SO is to use a custom spacer view in between:
`let space = UIBarButtonItem(barButtonSystemItem: .fixedSpace...`
I did not know about the `fixedSpace` API which might save us from all the logic in this file. I quite like this approach. If you agree and you don't have another specific one in mind, I would like to try it out and remove that file if it works.

Thanks for the feedback, I have found out about the latter API thanks to it!

Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Reader/Tab Navigation/DoubleBarControlConfigurer.swift]: I feel like we're getting sidetracked by an an issue that is unrelated to this project and isn't necessary to fix with it. A few minor sizing/spacing oddities clearly isn't causing any issues (and no one has noticed 😄 ).

 My suggestion is this:
- Let's not worry about fixing these issues right now, and revert to using the "normal" way of adding nav bar buttons (`navigationItem.rightBarButtonItems`).
- If you don't mind, please create a GitHub issue in the WP repo for the Reader nav bar issues, describing the problem and your research/findings. The screenshots you created for Matt Miklic would be helpful too.
- When this project is finished, we can add information to that issue for the Notifications nav bar (if the issue is still the same).

File: WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift (hunk 1 above)
Comment (ScoutHarris) [reply WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift:51]: Nice catch! That's been there a while too. 😄

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift (hunk 3 above)
Comment (ScoutHarris) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:1661]: Oh nice one!

**Review by ScoutHarris — CHANGES_REQUESTED** [review]
Re conversation about nav bar button fun.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift [hunk 6]
```diff
@@ -92,10 +92,33 @@ class NotificationsViewController: UITableViewController, UIViewControllerRestor
     }()
 
     /// Notification Settings button
-    lazy var notificationSettingsButton: UIBarButtonItem = {
-        let button = UIBarButtonItem(image: .gridicon(.cog), style: .plain, target: self, action: #selector(showNotificationSettings))
-        button.accessibilityLabel = NSLocalizedString("Notification Settings", comment: "Link to Notification Settings section")
-        return button
+    lazy var settingsBarButtonItem: UIBarButtonItem = {
```
Comment (ScoutHarris) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift]: Can you set this as `private` please? I know it wasn't before, but it should have been. As well as the new `markAllAsReadBarButtonItem` below.

**Review by ScoutHarris — APPROVED** [review]
Hey Alp.

I have one request. Otherwise this looks good! 🎉 

Just let me know when that change is made and I will merge this PR. Thanks!

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift (hunk 6 above)
Comment (alpavanoglu) [reply WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift]: Oh no that's a good point. Updating right away

**Pull request merged**